pkgs/development/tools/analysis/binlore/default.nix

```diff
@@ -56,58 +56,160 @@ let
+    uses a chain of two or more derivations where the output of one
+    is the source for the next. See resholve for an example.)
+  */
   collect = { lore ? loreDef, drvs, strip ? [ ] }: (runCommand "more-binlore" { } ''
     mkdir $out
     for lorefile in ${toString lore.types}; do
       cat ${lib.concatMapStrings (x: x + "/$lorefile ") (map (make lore) (map lib.getBin (builtins.filter lib.isDerivation drvs)))} > $out/$lorefile
-      substituteInPlace $out/$lorefile ${lib.concatMapStrings (x: "--replace '${x}/' '' ") strip}
+      substituteInPlace $out/$lorefile ${lib.concatMapStrings (x: "--replace-quiet '${x}/' '' ") strip}
     done
   '');
-  # TODO: echo for debug, can be removed at some point
+
+  /*
+    Output a directory containing lore for a single drv.
+
+    This produces lore for the derivation (via lore.callback) and
+    appends any lore that the derivation itself wrote to nix-support
+    or which was overridden in drv.binlore (passthru).
+
+    Since the last entry wins, the effective priority is:
+    drv.binlore > $drv/nix-support > lore generated here by callback
+  */
   make = lore: drv: runCommand "${drv.name}-binlore" {
-      identifier = drv.name;
       drv = drv;
     } (''
     mkdir $out
     touch $out/{${builtins.concatStringsSep "," lore.types}}
 
-    ${lore.callback lore drv overrides}
+    ${lore.callback lore drv}
+    '' +
+    # append lore from package's $out and drv.binlore (last entry wins)
+    ''
+    for lore_type in ${builtins.toString lore.types}; do
+      if [[ -f "${drv}/nix-support/$lore_type" ]]; then
+        cat "${drv}/nix-support/$lore_type" >> "$out/$lore_type"
+      fi
+      '' + lib.optionalString (builtins.hasAttr "binlore" drv) ''
+      if [[ -f "${drv.binlore}/$lore_type" ]]; then
+        cat "${drv.binlore}/$lore_type" >> "$out/$lore_type"
+      fi
+      '' + ''
+    done
 
     echo binlore for $drv written to $out
   '');
+
+  /*
+    Utility function for creating override lore for drv.
+
+    In the normal case, we attach this lore to `drv.passthru.binlore`.
+
+    > *Note*: We can reconsider the passthru attr name if someone adds
+    > a new lore provider. We settled on `.binlore` for now to make it
+    > easier for people to figure out what this is for.
+
+    The lore argument should be a Shell script (string) that generates
+    the necessary lore. You can use arbitrary Shell, but this function
+    includes a shell DSL you can use to declare/generate lore in most
+    cases. It has the following functions:
+
+    - `execer <verdict> [<path>...]`
+    - `wrapper <wrapper_path> <original_path>`
+
+    Writing every override explicitly in a Nix list would be tedious
+    for large packages, but this small shell DSL enables us to express
+    many overrides efficiently via pathname expansion/globbing.
+
+    Here's a very general example of both functions:
+
+    passthru.binlore = (binlore.synthesize finalAttrs.finalPackage ''
```

Comment:
Do you plan to support multiple outputs? For example, `glib` has programs in `out`, `bin` and `dev`.

Reply:
Good point/question, though I may need some advice.

I may be misunderstanding since I haven't intentionally crossed this bridge, but some naive play from CLI suggests passthru is reachable from each output, so I guess all we'd _have_ to do is make sure it's possible to generate one lore attr covering all of the outputs.

I guess I can imagine several approaches, but the best are probably:
1. Don't try to do the current convenience bit where the API cds into the output directory to expand/test the paths--make the call-sites explicitly inline the relevant output and do the expansion (and path testing) from there. Script format gets a little less efficient, but for most uses it'd be roughly balanced by dropping the 1st arg.
2. Accept a list instead of a single drv, try each path on every output in the list and only error if there's no match at all. Keeps the script format efficient but the implementation will get more complex in order to try expanding and testing each path on each drv and failing if a particular path didn't exist in any of them.
3. Basically the above, but auto-enumerate all of the outputs and try them all. This would keep the API simpler, but I guess on the fringe it'll force Nix to realize some outputs it wouldn't otherwise need if the lore override isn't cached.

I was imagining the second at first, but thinking it out a few steps has me leaning towards the first. WDYT? :)

Reply:
`passthru` is indeed intended to be added to all outputs, and it does not support customization per individual output.
Adding that is somewhat non-trivial, and `mkDerivation` needs to ditch `extendDerivation` for something more capable and integrated, which is arguably a good thing to do anyway. (Extending must never be done externally like that, but be handled by the abstraction that manages the fixpoint, which in this case is `mkDerivation` itself.)

Reply:
I explored the first approach, but I didn't like how verbose it was in `finalAttrs.finalPackage`  cases (and that it'd pull in extra dependencies unnecessarily in some cases).

Latest push adopts a 4th approach here that required less all-around change. `passthru.binlore` is now an attrset namespaced by output name. 

While there are no cases that need this behavior in the current set, I did test this out with a smaller change on top of nixpkgs-unstable by adding lore overrides for both glib.bin and glib.dev. I can push that up on a separate branch if anyone needs ~proof :)

Tentatively marking resolved, but I'm open to re-litigating the approach.